rust/src/lib/nm/query_apply/apply.rs

```diff
@@ -459,22 +466,38 @@ fn gen_nm_conn_need_to_deactivate_first(
     ret
 }
 
-fn check_nm_version(nm_api: &NmApi) {
-    if let Ok(versions) = nm_api.version().map(|ver_str| {
-        ver_str
+fn check_nm_version(nm_api: &NmApi, route_remove_needs_deactivate: &mut bool) {
+    let mut nm_ver_encoded = 0;
+
+    if let Ok(ver_info) = nm_api.version_info() {
+        nm_ver_encoded = ver_info.version_encoded();
+        *route_remove_needs_deactivate = !ver_info
+            .has_capability(NmVersionInfo::CAPABILITY_SYNC_ROUTE_WITH_TABLE);
+    } else if let Ok(version) = nm_api.version() {
+        // VersionInfo was added to NM 1.42. As we support 1.40+, fallback to
+        // parsing from Version string if VersionInfo is not available.
+        if let Ok(ver_arr) = version
+            .as_str()
             .split('.')
-            .map(|v| v.parse::<i32>().unwrap_or_default())
-            .collect::<Vec<i32>>()
-    }) {
-        if let (Some(major), Some(minor)) = (versions.first(), versions.get(1))
+            .take(2)
+            .map(|v| v.parse::<u32>())
+            .collect::<Result<Vec<u32>, _>>()
         {
-            if *major < 1 || *minor < 40 {
-                log::warn!(
-                    "Unsupported NetworkManager version {major}.{minor}, \
-                    expecting >= 1.40"
-                );
+            if ver_arr.len() == 2 {
+                nm_ver_encoded = ver_arr[0] << 16 | ver_arr[1] << 8;
             }
         }
+        *route_remove_needs_deactivate = true;
+    }
+
+    if nm_ver_encoded == 0 {
+        log::warn!("Unknown NetworkManager version, expecting >= 1.40");
+    } else if nm_ver_encoded < (1 << 16 | 40 << 8) {
```

Comment:
Ssome developer are using (major * 1000000 + minor * 1000 + micro), so this 16 and 8 is magic number to them.

If you want to improve old `*major < 1 || *minor < 40`. Please do something like:

```
struct NmVersion {
   major: u8
   minor: u8,
   micor: u8,
}

impl Ord for NmVersion {
}

impl TryFrom<&str> for NmVersion {
}
```

Reply:
Done

Reply:
I'm thinking that, instead of having structs NmVersion and NmVersion info, we could have only NmVersion instead. This NmVersion would hold the capabilities too. I would `NmApi::version_info()` and have only `NmApi::version()`. Then, `NmApi::version()` will try to fetch `VersionInfo` and, if fails, `Version`.

What do you think?